closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -469,39 +471,50 @@ int decryptFinal(byte[] in, int inOfs, int len,
                      byte[] out, int outOfs)
         throws IllegalBlockSizeException, AEADBadTagException,
         ShortBufferException {
+        // avoid any race conditions in OpenSSL when the key/iv are used concurrently
+        byte[] keyCopy = key.clone();
+        byte[] ivCopy = iv.clone();
```

Comment:
These copies may not be consistent, synchronization may be required.

Reply:
Can you please elaborate about that? Why can these copies be inconsistent?

Reply:
Those clones were in #59  that was accepted.

Reply:
If `key` and `iv` can be modified concurrently, then they may be updated (by other threads) in between the two clone operations such that the new values refer to arrays of incompatible lengths, for example.

I'm a little surprised that such concurrent use is possible via the crypto API. It would be helpful to have a fuller explanation about what sorts of concurrent uses are possible and which ones cause the issue this is intended to address.

Reply:
I see, so would you suggest I put the clone operations in the synchronized statement? I think to also resolve the comment below this (about `ibuffer` being modified before the if statements), I would need to change the code to the following:

```java
synchronized (this) {
    // avoid any race conditions in OpenSSL when the key/iv are used concurrently
    byte[] keyCopy = key.clone();
    byte[] ivCopy = iv.clone();
    int bufSize = ibuffer.size();

    if (len < (tagLenBytes - bufSize)) {
        throw new AEADBadTagException("Input too short - need tag");
    }

    if (len > (MAX_BUF_SIZE - bufSize)) {
        throw new ProviderException("SunJCE provider only supports "
            + "a positive input size up to " + MAX_BUF_SIZE + " bytes");
    }

    if (out.length - outOfs < len  + bufSize - tagLenBytes) {
        throw new ShortBufferException("Output buffer too small");
    }

    if (len > 0) {
        ibuffer.write(in, inOfs, len);
    }

    // refresh 'in' to all buffered-up bytes
    in = ibuffer.toByteArray();
    ibuffer.reset();
}
```

Do you think this makes sense?

Reply:
I don't think that will completely solve the problem; from my read of this and the non-native implementation, it's not reasonable for two (or more) threads to share a single `FeedbackCipher` object and thus sharing of `NativeGaloisCounterMode` objects is not well defined.

One thread might use invoke `encrypt(...)` one or more times before calling `encryptFinal(...)` while a second thread might just call the latter. If the second thread's call to `encryptFinal(...)` occurs after some of the first thread's calls to `encrypt(...)` but before its call to `encryptFinal(...)`, the second thread will erroneously use the data stored in `ibuffer_enc` by the first thread. No changes to `encryptFinal(...)` alone will correct that. I conclude that concurrent use of `FeedbackCipher` objects is not expected: thus the proposed changes don't address the fundamental issue (it could be that the test scenarios is unreasonable).

I repeat my request for more information about the way failures arise so that we may more fully understand the bigger picture.

Reply:
That is true, GCM Cipher objects are not supposed to be used concurrently. If used concurrently, the result is most probably going to be nonsense. However, this is not the bug that we aim to fix in this PR. This PR is fixing the native code crash that is happening. Adding this synchronization makes the native code not crash but give you an error instead. This (an error) is the expected behaviour if a user uses the same cipher object concurrently.

In terms of how the crash is happening, as per the [issue description and discussion](https://jazz103.hursley.ibm.com:9443/jazz/web/projects/JTC-JAT#action=com.ibm.team.workitem.viewWorkItem&id=145527), it seems that its happening when the SSL connection is being forcefully closed. Java is trying to finalize the GCM cipher before closing. When looking at the native code, Jason pointed out that the keyLen and ivLen are wrong

Reply:
Ok now that I better understand the concurrent interactions, I think we need `synchronized` blocks for all access of fields that require more consistency that we currently have. So I think that means `decryptFinal()` needs to be something like this:
```
    int decryptFinal(byte[] in, int inOfs, int len,
                     byte[] out, int outOfs)
        throws IllegalBlockSizeException, AEADBadTagException,
        ShortBufferException {
        if (len < 0) {
            throw new ProviderException("Input length is negative");
        }

        final int bufSize;
        final byte[] keyCopy;
        final byte[] ivCopy;
        final byte[] aad;
        // need to synchronize this block as if there are concurrent calls to decryptFinal
        // the shared variables aadBuffer and ibuffer may corrupt input and cause a segfault
        // in openssl code
        synchronized (this) {
            bufSize = ibuffer.size();
            if (len > 0) {
                ibuffer.write(in, inOfs, len);
            }

            if (len < (tagLenBytes - bufSize)) {
                throw new AEADBadTagException("Input too short - need tag");
            }

            if (len > (MAX_BUF_SIZE - bufSize)) {
                throw new ProviderException("SunJCE provider only supports "
                    + "a positive input size up to " + MAX_BUF_SIZE + " bytes");
            }

            if ((out.length - outOfs) < (len + bufSize - tagLenBytes)) {
                throw new ShortBufferException("Output buffer too small");
            }

            // refresh 'in' to all buffered-up bytes
            in = ibuffer.toByteArray();
            ibuffer.reset();

            // avoid any race conditions in OpenSSL when the key/iv are used concurrently
            keyCopy = key.clone();
            ivCopy = iv.clone();
            aad = ((aadBuffer == null) || (aadBuffer.size() == 0))
                    ? emptyAAD : aadBuffer.toByteArray();
            aadBuffer = null;
        }

        int ret = nativeCrypto.GCMDecrypt(keyCopy, keyCopy.length,
                ivCopy, ivCopy.length,
                in, 0, in.length,
                out, outOfs,
                aad, aad.length, tagLenBytes);

        if (ret == -2) {
            throw new AEADBadTagException("Tag mismatch!");
        } else if (ret == -1) {
            throw new ProviderException("Error in Native GaloisCounterMode");
        }

        return ret;
    }
```
Notice
* the ordering avoids cloning arrays until after most exceptional situations are ruled out
* access to `aadBuffer` is in the `synchronized` block

Similar changes are required to `encryptFinal` and all methods that access or modify relevant state.

Reply:
@keithc-ca Okay, I changed both functions as suggested above and addressed all the comments you have below. Please let me know if there's anything else.